usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +519,467 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+
+done:
+
+    if (blob_attr)
+        free(blob_attr);
+
+    return ret;
+}
+
+/**
+ * Return true if PKEY_MODE DISABLED is set in the token specific
+ * config file, false otherwise.
+ */
+static CK_BBOOL ep11tok_pkey_option_disabled(STDLL_TokData_t *tokdata)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+
+    if (ep11_data->pkey_mode == PKEY_MODE_DISABLED)
+        return CK_TRUE;
+
+    return CK_FALSE;
+}
+
+/**
+ * Return true, if the given key obj has a valid protected key, i.e. its
+ * verification pattern matches the one of the current master key.
+ */
+static CK_BBOOL ep11tok_pkey_is_valid(STDLL_TokData_t *tokdata, OBJECT *key_obj)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_ATTRIBUTE *pkey_attr = NULL;
+    int vp_offset;
+
+    if (template_attribute_get_non_empty(key_obj->template, CKA_IBM_OPAQUE_PKEY,
+                                         &pkey_attr) == CKR_OK) {
+        if (pkey_attr->ulValueLen >= 16 + MK_VP_LENGTH) {
+            vp_offset = pkey_attr->ulValueLen - MK_VP_LENGTH;
+            if (memcmp(pkey_attr->pValue + vp_offset,
+                       &ep11_data->pkey_mk_vp,
+                       MK_VP_LENGTH) == 0) {
+                return CK_TRUE;
+            }
+        }
+    }
+
+    return CK_FALSE;
+}
+
+/**
+ * Create a new protected key for the given key obj and update attribute
+ * CKA_IBM_OPAQUE with the new pkey.
+ */
+static CK_RV ep11tok_pkey_update(STDLL_TokData_t *tokdata, OBJECT *key_obj)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_ATTRIBUTE *skey_attr = NULL;
+    CK_ATTRIBUTE *pkey_attr = NULL;
+    CK_RV ret;
+    int vp_offset;
+
+    /* Get secure key from obj */
+    if (template_attribute_get_non_empty(key_obj->template, CKA_IBM_OPAQUE,
+                                         &skey_attr) != CKR_OK) {
+        TRACE_ERROR("This key has no blob: should not occur!\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Transform the secure key into a protected key */
+    ret = ep11tok_pkey_skey2pkey(tokdata, skey_attr, &pkey_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("protected key creation failed with rc=0x%lx\n",ret);
+        goto done;
+    }
+
+    /* Check if the new pkey's verification pattern matches the one in
+     * ep11_data. This should always be the case, because we just
+     * created the pkey with the current MK. */
+    vp_offset = pkey_attr->ulValueLen - MK_VP_LENGTH;
+    if (memcmp(&ep11_data->pkey_mk_vp, pkey_attr->pValue + vp_offset,
+               MK_VP_LENGTH) != 0) {
+        TRACE_ERROR("vp of this pkey does not match with the one in ep11_data (should not occur)\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Now update the key obj. If it's a token obj, it will be also updated
+     * in the repository. */
+    ret = pkey_update_and_save(tokdata, key_obj, pkey_attr);
+    if (ret != CKR_OK) {
```

Comment:
Because pkey_update_and_save unlocks and relocks the object, it might get concurrently updated by other threads at this place. 

Since you just return from this function after returning from pkey_update_and_save it should be save for this function, but how about the callers?

Reply:
See rework in all routines calling ep11tok_pkey_check. We fall back into the ep11 path only if no attempt was made to create a pkey, i.e. no unlock/relock occurred.